// .ignored/eslint-plugin-sql/dist/rules/format.js
import { createRule } from '../factories/createRule.js';
import { dropBaseIndent } from '../utilities/dropBaseIndent.js';
import { isSqlQuery } from '../utilities/isSqlQuery.js';
import { AST_NODE_TYPES } from '@typescript-eslint/utils';
import { format } from 'sql-formatter';
const padIndent = (subject, indentString, indentCount) => {
    return subject
        .split('\n')
        .map((line) => {
        return line.length > 0 ? indentString.repeat(indentCount) + line : line;
    })
        .join('\n');
};
const detectIndentation = (subject) => {
    for (const line of subject.split('\n')) {
        if (line.trim().length > 0) {
            const match = line.match(/^(\t+|\u0020+)/u);
            if (match) {
                const leadingWhitespace = match[1];
                return {
                    char: leadingWhitespace[0] === '\t' ? '\t' : ' ',
                    count: leadingWhitespace.length,
                };
            }
            return { char: ' ', count: 0 };
        }
    }
    return { char: ' ', count: 0 };
};
const getClosingIndentation = (subject) => {
    const lines = subject.split('\n');
    const lastLine = lines[lines.length - 1];
    if (lastLine && lastLine.trim().length === 0 && lastLine.length > 0) {
        const char = lastLine[0] === '\t' ? '\t' : ' ';
        return { char, count: lastLine.length };
    }
    return { char: ' ', count: 0 };
};
const calculateIndentConfig = (templateRaw, useTabs, tabWidth) => {
    // Detect the original indentation style from the template
    let detectedIndent = detectIndentation(templateRaw);
    // If no meaningful content, try to detect from the closing line
    const hasNoContent = templateRaw.search(/\S/u) === -1;
    if (hasNoContent) {
        detectedIndent = getClosingIndentation(templateRaw);
    }
    // Determine the indent character to use:
    // 1. If useTabs is explicitly set, use that
    // 2. Otherwise, use the detected indentation from the original template
    const indentChar = useTabs === undefined ? detectedIndent.char : useTabs ? '\t' : ' ';
    // Calculate the indent count for the SQL content
    let contentCount;
    if (hasNoContent) {
        // No meaningful content - use closing line indentation + 1 level
        contentCount =
            indentChar === '\t'
                ? detectedIndent.count + 1
                : detectedIndent.count + tabWidth;
    }
    else if (templateRaw.search(/\S/u) === 0) {
        // Content starts immediately (no leading whitespace)
        contentCount = indentChar === '\t' ? 1 : tabWidth;
    }
    else {
        // Use detected indentation
        contentCount = detectedIndent.count;
    }
    // Calculate closing backtick indentation (one level less than content)
    const closingCount = indentChar === '\t'
        ? Math.max(contentCount - 1, 0)
        : Math.max(contentCount - tabWidth, 0);
    return {
        char: indentChar,
        closingCount,
        contentCount,
    };
};
export const rule = createRule({
    create: (context) => {
        var _a, _b, _c, _d, _e, _f, _g, _h;
        // @ts-expect-error I am ont clear how to type this
        const placeholderRule = (_b = (_a = context.settings) === null || _a === void 0 ? void 0 : _a.sql) === null || _b === void 0 ? void 0 : _b.placeholderRule;
        const pluginOptions = ((_c = context.options) === null || _c === void 0 ? void 0 : _c[0]) || {
            ignoreExpressions: false,
            ignoreInline: true,
            ignoreStartWithNewLine: true,
            ignoreTagless: true,
            retainBaseIndent: true,
            sqlTag: 'sql',
        };
        const { ignoreExpressions, ignoreInline, ignoreStartWithNewLine, ignoreTagless, retainBaseIndent, sqlTag, } = pluginOptions;
        const tabWidth = (_f = (_e = (_d = context.options) === null || _d === void 0 ? void 0 : _d[1]) === null || _e === void 0 ? void 0 : _e.tabWidth) !== null && _f !== void 0 ? _f : 2;
        const useTabs = (_h = (_g = context.options) === null || _g === void 0 ? void 0 : _g[1]) === null || _h === void 0 ? void 0 : _h.useTabs;
        return {
            TemplateLiteral(node) {
                var _a, _b, _c, _d, _e, _f, _g, _h;
                const tagName = 
                // @ts-expect-error TODO
                (_e = (_b = (_a = node.parent.tag) === null || _a === void 0 ? void 0 : _a.name) !== null && _b !== void 0 ? _b : 
                // @ts-expect-error TODO
                (_d = (_c = node.parent.tag) === null || _c === void 0 ? void 0 : _c.object) === null || _d === void 0 ? void 0 : _d.name) !== null && _e !== void 0 ? _e : 
                // @ts-expect-error TODO
                (_h = (_g = (_f = node.parent.tag) === null || _f === void 0 ? void 0 : _f.callee) === null || _g === void 0 ? void 0 : _g.object) === null || _h === void 0 ? void 0 : _h.name;
                const sqlTagIsPresent = tagName === sqlTag;
                if (ignoreTagless && !sqlTagIsPresent) {
                    return;
                }
                if (ignoreExpressions && node.quasis.length !== 1) {
                    return;
                }
                const templateElement = node.quasis.find((quasi) => {
                    return quasi.type === AST_NODE_TYPES.TemplateElement;
                });
                if (!templateElement) {
                    return;
                }
                const magic = '"gajus-eslint-plugin-sql"';
                const literal = node.quasis
                    .map((quasi) => {
                    return quasi.value.raw;
                })
                    .join(magic);
                const indentConfig = calculateIndentConfig(literal, useTabs, tabWidth);
                if (!sqlTagIsPresent && !isSqlQuery(literal, placeholderRule)) {
                    return;
                }
                if (ignoreInline && !literal.includes('\n')) {
                    return;
                }
                let formatted;
                try {
                    formatted = format(literal, {
                        ...context.options[1],
                        tabWidth,
                    });
                }
                catch (error) {
                    context.report({
                        data: {
                            message: error instanceof Error ? error.message : 'Unknown parse error',
                        },
                        messageId: 'parseError',
                        node,
                    });
                    return;
                }
                if (ignoreStartWithNewLine &&
                    literal.startsWith('\n') &&
                    !formatted.startsWith('\n')) {
                    formatted = '\n' + formatted;
                }
                if (formatted.endsWith('\n\n')) {
                    formatted = formatted.replace(/\n\n$/u, '\n');
                }
                if (retainBaseIndent) {
                    formatted = padIndent(formatted, indentConfig.char, indentConfig.contentCount);
                    formatted +=
                        '\n' + indentConfig.char.repeat(indentConfig.closingCount);
                }
                else {
                    formatted = dropBaseIndent(formatted);
                    formatted += '\n';
                }
                if (formatted !== literal) {
                    context.report({
                        fix: (fixer) => {
                            let final = formatted;
                            const expressionCount = node.expressions.length;
                            let index = 0;
                            while (index <= expressionCount - 1) {
                                final = final.replace(magic, '${' +
                                    context.sourceCode.getText(node.expressions[index]) +
                                    '}');
                                index++;
                            }
                            return fixer.replaceTextRange([
                                node.quasis[0].range[0],
                                node.quasis[node.quasis.length - 1].range[1],
                            ], '`' + (final.startsWith('\n') ? final : '\n' + final) + '`');
                        },
                        messageId: 'format',
                        node,
                    });
                }
            },
        };
    },
    defaultOptions: [
        {
            ignoreExpressions: false,
            ignoreInline: true,
            ignoreStartWithNewLine: true,
            ignoreTagless: true,
            retainBaseIndent: true,
            sqlTag: 'sql',
        },
        {
            tabWidth: 2,
        },
    ],
    meta: {
        docs: {
            description: 'Matches queries in template literals. Warns when query formatting does not match the configured format (see Options).',
            url: 'https://github.com/gajus/eslint-plugin-sql#eslint-plugin-sql-rules-format',
        },
        fixable: 'code',
        messages: {
            format: 'Format the query',
            parseError: 'SQL parse error: {{message}}',
        },
        schema: [
            {
                additionalProperties: false,
                properties: {
                    ignoreExpressions: {
                        default: false,
                        type: 'boolean',
                    },
                    ignoreInline: {
                        default: true,
                        type: 'boolean',
                    },
                    ignoreStartWithNewLine: {
                        default: true,
                        type: 'boolean',
                    },
                    ignoreTagless: {
                        default: true,
                        type: 'boolean',
                    },
                    retainBaseIndent: {
                        default: true,
                        type: 'boolean',
                    },
                    sqlTag: {
                        default: 'sql',
                        type: 'string',
                    },
                },
                type: 'object',
            },
            {
                additionalProperties: false,
                properties: {
                    dataTypeCase: {
                        default: 'preserve',
                        enum: ['lower', 'upper', 'preserve'],
                        type: 'string',
                    },
                    denseOperators: {
                        default: false,
                        type: 'boolean',
                    },
                    functionCase: {
                        default: 'preserve',
                        enum: ['lower', 'upper', 'preserve'],
                        type: 'string',
                    },
                    identifierCase: {
                        default: 'preserve',
                        enum: ['lower', 'upper', 'preserve'],
                        type: 'string',
                    },
                    keywordCase: {
                        default: 'preserve',
                        enum: ['lower', 'upper', 'preserve'],
                        type: 'string',
                    },
                    language: {
                        default: 'sql',
                        type: 'string',
                    },
                    paramTypes: {
                        additionalProperties: false,
                        properties: {
                            custom: {
                                items: {
                                    additionalProperties: false,
                                    properties: {
                                        regex: {
                                            type: 'string',
                                        },
                                    },
                                    required: ['regex'],
                                    type: 'object',
                                },
                                type: 'array',
                            },
                            named: {
                                items: {
                                    enum: [':', '@', '$'],
                                    type: 'string',
                                },
                                type: 'array',
                            },
                            numbered: {
                                items: {
                                    enum: ['?', ':', '$'],
                                    type: 'string',
                                },
                                type: 'array',
                            },
                            positional: {
                                type: 'boolean',
                            },
                            quoted: {
                                items: {
                                    enum: [':', '@', '$'],
                                    type: 'string',
                                },
                                type: 'array',
                            },
                        },
                        type: 'object',
                    },
                    tabWidth: {
                        default: 2,
                        type: 'number',
                    },
                    useTabs: {
                        default: false,
                        type: 'boolean',
                    },
                },
                type: 'object',
            },
        ],
        type: 'suggestion',
    },
    name: 'format',
});
